docs(headless,headless-react): refactor typedoc to central location

jsmith2-coveo opened the pull request:
Many of the files for the generation of typedoc sites are duplicated between the headless and headless-react packages. Combining them under a centralized shared typedoc location, and referencing that during the corresponding builds would be more efficient.

jpmarceau reviewed (approved):
👌 
Looks great to me, but I'll be curious to hear the DX team on whether a new sub-package is an issue for them.

louis-bompart commented on packages/shared-typedoc/package.json:
Should be private!

louis-bompart reviewed (changes requested):
While the refactor is good, how we referenced the refactored files ain't.
We should use package specifiers to do so;
(i.e. import from or extends from shared-typedoc package name)

louis-bompart commented on packages/shared-typedoc/tsconfig.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "extends": "../headless-react/tsconfig.json",
```
Would it be possible to not have this extends; I think it does not make sense architecturally

jsmith2-coveo replied:
You're right, I removed the extends and moved some of the required functionality directly into the `tsconfig.json`

jsmith2-coveo commented:
I added the use of package specifiers where I could, and adjusted the location of the config files to be localized to `headless` and `headless-react` packages respectfully. I found that keeping the configs in the shared-typedoc location when using package reference was causing issues. 

I believe the changes have further improved the restructuring approach to Typedoc, and would be happy to adjust accordingly for any further feedback! 😄

alexprudhomme commented on packages/shared-typedoc/lib/formatTypeDocToolbar.d.ts:
```diff
@@ -0,0 +1 @@
+export declare function formatTypeDocToolbar(): void;
```
I don't think any of those `.d.ts` files are necessary. I would remove them all.

alexprudhomme replied:
They should only be present in the `dist` folder after building and not in the `lib` folder

jsmith2-coveo replied:
Good call, I didn't catch these remnants from the early build process

alexprudhomme commented on packages/shared-typedoc/package.json:
```diff
@@ -0,0 +1,19 @@
+{
```
You need to add a project.json file for this package to make sure NX runs the build properly at the right time.

It would look something like this : 

```
{
  "name": "shared-typedoc",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "targets": {
    "build": {
      "executor": "nx:run-commands",
      "outputs": ["{projectRoot}/dist"],
      "cache": true,
      "options": {
        "commands": [
          "tsc -p tsconfig.json",
        ],
        "cwd": "{projectRoot}"
      }
    }
  }
}
```

alexprudhomme commented on packages/shared-typedoc/package.json:
```diff
@@ -0,0 +1,19 @@
+    "tsconfig.json"
+  ],
+  "scripts": {
+    "build": "npx tsc -p tsconfig.json"
```
After adding the `project.json` file, you can change this line to this.
```suggestion
    "build": "nx build"
```

jsmith2-coveo commented on package.json:
```diff
@@ -94,6 +94,7 @@
     "packages/atomic-hosted-page",
     "packages/atomic-react",
     "packages/atomic-angular",
+    "packages/shared-typedoc",
```
I've renamed the package to "documentation"!

louis-bompart reviewed (approved):
Not an easy one, good job!
